analyzers/src/SonarAnalyzer.CSharp/Rules/TimestampsShouldNotBeUsedAsPrimaryKeys.cs

```diff
@@ -0,0 +1,106 @@
+                // This results in a couple of FNs, but those scenarios are very rare.
+                var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                var className = classDeclaration.Identifier.ValueText;
+                var keyProperties = classDeclaration.Members
+                    .OfType<PropertyDeclarationSyntax>()
+                    .Where(x => IsPublicReadWriteProperty(x) && IsTemporalType(x) && IsKeyProperty(x, className));
+
+                foreach (var keyProperty in keyProperties)
+                {
+                    c.ReportIssue(Diagnostic.Create(Rule, keyProperty.Type.GetLocation(), keyProperty.Type.GetName()));
+                }
+            }, SyntaxKind.ClassDeclaration);
+        });
+
+    private static bool IsReferencingEntityFramework(Compilation compilation) =>
+        compilation.GetTypeByMetadataName(KnownType.Microsoft_EntityFrameworkCore_DbContext) is not null
+        || compilation.GetTypeByMetadataName(KnownType.Microsoft_EntityFramework_DbContext) is not null;
+
+    private static bool IsPublicReadWriteProperty(PropertyDeclarationSyntax property) =>
+        property.Modifiers.Any(x => x.IsKind(SyntaxKind.PublicKeyword))
+        && property.AccessorList is not null
+        && property.AccessorList.Accessors.Any(x => x.Keyword.IsKind(SyntaxKind.GetKeyword))
+        && property.AccessorList.Accessors.Any(x => x.Keyword.IsKind(SyntaxKind.SetKeyword));
+
+    private static bool IsKeyProperty(PropertyDeclarationSyntax property, string className)
+    {
+        var propertyName = property.Identifier.ValueText;
+        return propertyName.Equals("Id", StringComparison.InvariantCultureIgnoreCase)
+            || propertyName.Equals($"{className}Id", StringComparison.InvariantCultureIgnoreCase)
+            || HasKeyAttribute(property);
+    }
+
+    private static bool HasKeyAttribute(PropertyDeclarationSyntax property) =>
+        property.AttributeLists
+            .SelectMany(x => x.Attributes)
+            .Any(x => MatchesAttributeName(x, KeyAttributeTypeNames));
+
+    private static string[] TypeNamesForAttribute(KnownType attributeType) => new[]
+    {
+        attributeType.TypeName,
+        RemoveFromEnd(attributeType.TypeName, "Attribute"),
+        attributeType.FullName,
+        RemoveFromEnd(attributeType.FullName, "Attribute"),
+    };
+
+    private static string RemoveFromEnd(string text, string subtextFromEnd) =>
+        text.Substring(0, text.LastIndexOf(subtextFromEnd));
+
+    private static bool IsTemporalType(PropertyDeclarationSyntax property) =>
+        Array.Exists(TemporalTypes, x => property.Type.NameIs(x.TypeName) || property.Type.NameIs(x.FullName));
+
+    private static bool MatchesAttributeName(AttributeSyntax attribute, string[] candidates) =>
+        Array.Exists(candidates, x => attribute.NameIs(x));
```

Comment:
Wouldn't it be shorter and more understandable with `candidates.Any(x => attribute.NameIs(x));`
If you agree to using Any instead, I suggest doing the change for the IsTemporalType too.